src/werkzeug/test.py

```diff
@@ -467,162 +467,147 @@ def base_url(self, value):
 
-    content_type = property(
-        _get_content_type,
-        _set_content_type,
-        doc="""The content type for the request.  Reflected from and to
-        the :attr:`headers`.  Do not set if you set :attr:`files` or
-        :attr:`form` for auto detection.""",
-    )
-    del _get_content_type, _set_content_type
-
-    def _get_content_length(self):
-        return self.headers.get("Content-Length", type=int)
+    @property
+    def mimetype(self):
+        """The mimetype (content type without charset etc.)
 
-    def _get_mimetype(self):
+        .. versionadded:: 0.14
+        """
         ct = self.content_type
         if ct:
             return ct.split(";")[0].strip()
+        return None
 
-    def _set_mimetype(self, value):
+    @mimetype.setter
+    def mimetype(self, value):
         self.content_type = get_content_type(value, self.charset)
 
-    def _get_mimetype_params(self):
+    @property
+    def mimetype_params(self):
+        """ The mimetype parameters as dict.  For example if the
+        content type is ``text/html; charset=utf-8`` the params would be
+        ``{'charset': 'utf-8'}``.
+
+        .. versionadded:: 0.14
+        """
+
         def on_update(d):
             self.headers["Content-Type"] = dump_options_header(self.mimetype, d)
 
         d = parse_options_header(self.headers.get("content-type", ""))[1]
         return CallbackDict(d, on_update)
 
-    mimetype = property(
-        _get_mimetype,
-        _set_mimetype,
-        doc="""The mimetype (content type without charset etc.)
-
-        .. versionadded:: 0.14
-    """,
-    )
-    mimetype_params = property(
-        _get_mimetype_params,
-        doc=""" The mimetype parameters as dict.  For example if the
-        content type is ``text/html; charset=utf-8`` the params would be
-        ``{'charset': 'utf-8'}``.
-
-        .. versionadded:: 0.14
-        """,
-    )
-    del _get_mimetype, _set_mimetype, _get_mimetype_params
+    @property
+    def content_length(self):
+        """The content length as integer.  Reflected from and to the
+        :attr:`headers`.  Do not set if you set :attr:`files` or
+        :attr:`form` for auto detection.
+        """
+        return self.headers.get("Content-Length", type=int)
 
-    def _set_content_length(self, value):
+    @content_length.setter
+    def content_length(self, value):
         if value is None:
             self.headers.pop("Content-Length", None)
         else:
             self.headers["Content-Length"] = str(value)
 
-    content_length = property(
-        _get_content_length,
-        _set_content_length,
-        doc="""The content length as integer.  Reflected from and to the
-        :attr:`headers`.  Do not set if you set :attr:`files` or
-        :attr:`form` for auto detection.""",
-    )
-    del _get_content_length, _set_content_length
-
-    def form_property(name, storage, doc):  # noqa: B902
-        key = "_" + name
-
```

Comment:
Here you're flattening out a function that returns a descriptor and is invoked twice into duplicated copies. This is increasing code duplication, although eliminating a level of abstraction.

I would err on the side of not replacing this one.

Reply:
Good catch, but after looking at both I think I prefer the new version. I'll add a comment to both that they're related, but I don't expect these properties to change much.

Reply:
I think I prefer a hybrid of both approaches, I'll extract the behavior to some common private methods.